src/main/java/com/faforever/client/FafClientApplication.java

```diff
@@ -120,5 +122,30 @@ private void showMainWindow() {
   public void stop() throws Exception {
     applicationContext.close();
     super.stop();
+
+    Thread timeoutThread = new Thread(() -> {
+      try {
+        Thread.sleep(Duration.ofSeconds(30).toMillis());
+      } catch (InterruptedException e) {
+      }
+
+      Thread.getAllStackTraces()
+          .entrySet()
+          .stream()
+          .filter(e -> !e.getKey().isDaemon())
+          .forEach(e -> {
+            log.error("Non daemon Thread \"{}\" (id: {}) still active in state: {}", e.getKey().getName(), e.getKey().getId(), e.getKey().getState());
+            log.error("Stacktrace of thread {}:\n{}", e.getKey().getName(), Arrays.stream(e.getValue()).map(Object::toString).collect(Collectors.joining("\n")));
+          });
```

Comment:
Maybe it should collect the non-daemon threads and if the size is 0 it should return normally. I know that daemon threads are supposed to terminate while in-process but maybe they get interrupted first and then you could just return.

Reply:
Yeah maybe could be. Tho chances are small that happens i think. Even tho not sure if it happens at all

Reply:
done